Niamh25/doc 5383/corda 5 advanced utxo ledger extensions (#1385)
corda/corda-docs-portal: content/en/platform/corda/5.0/developing-applications/getting-started/utxo-advanced-ledger-extensions/ledger-extensions-api-reference.md

---
date: '2023-06-01'
title: "Ledger Extensions API Reference"
project: corda
version: 'Corda 5.0'
menu:
  corda5:
    identifier: corda5-utxo-ledger-extensions-api-reference
    parent: corda5-utxo-advanced-ledger-extensions
    weight: 4600
section_menu: corda5
---

> # Advanced Contract Design

Reviewer: The heading weights still aren't right - the "Advanced Contract Design" paragraph is a not on the following, it's not the heading for the whole page.
There should be another heading like "Advanced Ledger Types" or something.

Reviewer 2: Updated to read:

# Ledger Extensions API Reference

## Advanced Contract Design

...

## Advanced Ledger Types

### Base API

> # Advanced Contract Design

Reviewer 3: ```suggestion
# Ledger Extensions API Reference
```

Reviewer 2: Have updated the headings based on Christian's suggestion to read:

# Ledger Extensions API Reference

## Advanced Contract Design

...
## Advanced Ledger Types

### Base API

## Chainable API

Module: chainable

Package: com.r3.corda.ledger.utxo.chainable

The chainable API provides the component model for designing chainable states and contracts. Chainable states represent strictly linear state chains, where every state in the chain points to the previous state in the chain. This could be thought of as a similar concept to a blockchain, where each new block points to the previous block.

### Designing Chainable States

A chainable state can be implemented by implementing the `ChainableState<T>` interface; for example:

```kotlin
@BelongsToContract(ExampleChainableContract.class)
public final class ExampleChainableState extends ChainableState<ExampleChainableState> {
  @Nullable
  private final StaticPointer<ExampleChainableState> pointer;
  
  public ExampleChainableState(@NotNull final StaticPointer<ExampleChainableState> pointer) {
    this.pointer = pointer;
  }
  
  @Nullable 
  public StaticPointer<ExampleChainableState> getPreviousStatePointer() {
    return pointer;
  }
  
  @NotNull
  public List<PublicKey> getParticipants() {
    return List.of(...);
  }
}
```

### Designing Chainable Commands

> Chainable commands allows users to create, update, and delete chainable states.

Reviewer 3: ```suggestion
Chainable commands allow users to create, update, and delete chainable states.
```

Reviewer 2: Done.

> The `ChainableContractCreateCommand` creates new chainable states and will verify the following constraints:

Reviewer 2: Done (and all other instances).

* On chainable state(s) creating, at least one chainable state must be created.
* On chainable state(s) creating, the previous state pointer of every created chainable state must be null.

```kotlin
public final class Create extends ChainableContractCreateCommand<ExampleChainableState> {
  @NotNull
  public Class<ExampleChainableState> getContractStateType() {
    return ExampleChainableState.class;
  }
  
  @Override
  protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
    // Verify additional Create constraints
  }
}
```

> The `ChainableContractUpdateCommand` supports updating existing chainable states and will verify the following constraints:

Reviewer 2: Done

* On chainable state(s) updating, at least one chainable state must be consumed.
* On chainable state(s) updating, at least one chainable state must be created.
* On chainable state(s) updating, the previous state pointer of every created chainable state must not be null.
* On chainable state(s) updating, the previous state pointer of every created chainable state must be pointing to exactly one consumed chainable state, exclusively.

```kotlin
public final class Update extends ChainableContractUpdateCommand<ExampleChainableState> {
  @NotNull
  public Class<ExampleChainableState> getContractStateType() {
    return ExampleChainableState.class;
  }
  
  @Override
  protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {    
    // Verify additional Update constraints
  }
}
```

> The `ChainableContractDeleteCommand` supports deleting existing chainable states and will verify the following constraint:

Reviewer 2: Done

## Fungible API

Module: fungible

Package: com.r3.corda.ledger.utxo.fungible

The fungible API provides the component model for designing fungible states and contracts. Fungible states represent states that have a scalar numeric quantity, and can be split, merged and mutually exchanged with other fungible states of the same class. Fungible states represent the building blocks for states like tokens.

### Designing Fungible States

A fungible state can be implemented by implementing the `FungibleState<T>` interface; for example:

```kotlin
public final class ExampleFungibleState extends FungibleState<NumericDecimal> {  
  @NotNull
  private final NumericDecimal quantity;
  
  public ExampleFungibleState(@NotNull final NumericDecimal quantity) {
    this.quantity = quantity;
  }
  
  @NotNull
  public NumericDecimal getQuantity() {
    return quantity;
  }
  
  @NotNull
  public List<PublicKey> getParticipants() {
    return List.of(...);
  }
  
  @Override
  public boolean isFungibleWith(@NotNull final FungibleState<NumericDecimal> other) {
    return this == other || other instanceof ExampleFungibleState // && other fungibility rules.
  }
}
```

### Designing Fungible Commands

> Fungible commands allows users to create, update and delete fungible states.

Reviewer 2: Done

> The `FungibleContractCreateCommand` creates new fungible states and will verify the following constraints:

Reviewer 3: ```suggestion
The `FungibleContractCreateCommand` creates new fungible states and verifies the following constraints:
```

Reviewer 2: Done

* On fungible state(s) creating, at least one fungible state must be created.
* On fungible state(s) creating, the quantity of every created fungible state must be greater than zero.

```kotlin
public final class Create extends FungibleContractCreateCommand<ExampleFungibleState> {
  @NotNull
  public Class<ExampleFungibleState> getContractStateType() {
    return ExampleFungibleState.class;
  }
  
  @Override
  protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
    // Verify additional Create constraints
  }
}
```

> The `FungibleContractUpdateCommand` supports updating existing fungible states and will verify the following constraints:

Reviewer 2: Done

* On fungible state(s) updating, at least one fungible state must be consumed.
* On fungible state(s) updating, at least one fungible state must be created.
* On fungible state(s) updating, the quantity of every created fungible state must be greater than zero.
* On fungible state(s) updating, the sum of the unscaled values of the consumed states must be equal to the sum of the unscaled values of the created states.
* On fungible state(s) updating, the sum of the consumed states that are fungible with each other must be equal to the sum of the created states that are fungible with each other.

```kotlin
public final class Update extends FungibleContractUpdateCommand<ExampleFungibleState> {
  @NotNull
  public Class<ExampleFungibleState> getContractStateType() {
    return ExampleFungibleState.class;
  }
  
  @Override
  protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
    // Verify additional Update constraints
  }
}
```

> The `FungibleContractDeleteCommand` supports deleting existing fungible states and will verify the following constraints:

Reviewer 3: ```suggestion
The `FungibleContractDeleteCommand` supports deleting existing fungible states and verifies the following constraints:
```

Reviewer 2: Done

### Designing Identifiable States

An identifiable state can be implemented by implementing the `IdentifiableState` interface, for example:

```kotlin
public final class ExampleIdentifiableState extends IdentifiableState {
  @Nullable
  private final StateRef id;
  
  public ExampleIdentifiableState(@Nullable final StateRef id) {
    this.id = id;
  }
  
  @Nullable
  public StateRef getId() {
    return id;
  }
  
  @NotNull
  public List<PublicKey> getParticipants() {
    return List.of(...);
  }
}
```

### Designing Identifiable Commands

Identifiable commands support creating, updating and deleting identifiable states.
The `IdentifiableContractCreateCommand` supports creating new identifiable states and verifies the identifiable state(s) creation, at least one identifiable state must be created.

```kotlin
public final class Create extends IdentifiableContractCreateCommand<ExampleIdentifiableState> {
  @NotNull
  public Class<ExampleIdentifiableState> getContractStateType() {
    return ExampleIdentifiableState.class;
  }
  
  @Override
  protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
    // Verify additional Create constraints
  }
}
```

> The `IdentifiableContractUpdateCommand` updates existing identifiable states and will verify the following constraints:

Reviewer 2: Done

* On identifiable state(s) updating, at least one identifiable state must be consumed.
* On identifiable state(s) updating, at least one identifiable state must be created.

> * On identifiable state(s) updating, each created identifiable state's identifier must match one consumed identifiable state's state ref or identifier, exclusively.

Reviewer 3: ```suggestion
* On identifiable state(s) updating, each created identifiable state's identifier must match one consumed identifiable state's state reference or identifier, exclusively.
```

Reviewer 2: Done

    return ExampleIdentifiableState.class;
  }
  
  @Override
  protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
    // Verify additional Update constraints
  }
}
```

The `IdentifiableContractDeleteCommand` deletes existing identifiable states and verifies the identifiable state(s) deletion, at least one identifiable state must be consumed.

```kotlin
public final class Delete extends IdentifiableContractDeleteCommand<ExampleIdentifiableState> {
  @NotNull
  public Class<ExampleIdentifiableState> getContractStateType() {
    return ExampleIdentifiableState.class;
  }
  
  @Override
  protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
    // Verify additional Delete constraints
  }
}
```

### Designing Identifiable Contracts

An identifiable contract can be implemented by extending the `IdentifiableContract` class, for example:

```kotlin
public final class ExampleIdentifiableContract extends IdentifiableContract {
  @Override
  public List<Class<? extends IdentifiableContractCommand<?>>> getPermittedCommandTypes() {
    return List.of(Create.class, Update.class, Delete.class);
  }
}

> ```

Reviewer: There are sections missing on `IssuableState` and `OwnableState`.

Reviewer 2: Have added the material you provided.
